INTEXT-187 - Allow adding advice handler on AWS outbound channel adapter:

pckeyan opened the pull request:
1) Added Outbound S3 base type with a choice element for request-handler-advice-chain. Lines Starting 64-217
	2) Referred the base type in S3 Outboud adapter from Lines 219-230

Please review my changes.

pckeyan commented:
I already signed the CLA agreement; Please review my changes.

artembilan commented:
Hi @pckeyan !

Thank you for the comming back and contribution, of course :smile: 

1. No reason to say about CLA each time. If that we'll ask you
2. Please, add some test-case on the matter. That is a good tone to follow with the Contribution Guideline...

artembilan commented:
The fix looks good, but would you mind taking care about it a bit farther:
1. I don't see reason to extract `outboundS3BaseType`. You can do the same directly on the `s3-outbound-channel-adapter`
2. Please, try to follow with the configuration for the `sqs-outbound-channel-adapter`. My main concern is around `<xsd:attributeGroup ref="integration:channelAdapterAttributes"/>` to cover `channel`, `id` and smart-lifecycle attributes.
3. And yes: test is important here! Just because having it before we haven't faced such an issue at all!
Even if it is a new feature from the Spring Integration Core, we should cover it with some test on the matter. You can find similar tests, e.g. in the `FileOutboundChannelAdapterParserTests-context.xml` of Spring Integration File module.

pckeyan commented:
Sure, I will add test case and incorporate your comments.

pckeyan commented:
I updated the review comments, Please review at your convenience.

artembilan commented:
Hello @pckeyan !

So, let me review this PR and maybe we will merge it soon and will celebrate your first commit :smile: 

The general comments, first of all:

1. For better readability it is good to have header of the first commit in length <= 50 symbols.
2. It's good practice to add a JIRA link to the commit to reference each other for traceability.
3. The new classes must be in the current year in the `@Copyright` section. As well as the affected must be updated to the current year.
4. Changes in classes must follow with the code style rules: https://github.com/spring-projects/spring-framework/wiki/Spring-Framework-Code-Style
5. Please, configure your IDE to add a blank line in the end of each affected file.
6. Please, try to follow with my comments in PRs. I have a lot of them to review, plus I should some other codding work. So, I might not remember everything what I asked or suggested. For example, I don't see changes according to the item # 2 in my previous comment.

These all for the future contributions. This one, as first, is OK and I'll just polish it on the matter and show here the file commit. 

From other side that is just a recommendation to save our and your time during the contribution process.

Right now let me go to the lines for more closer comments.

artembilan commented on .gitignore:
```diff
@@ -8,4 +8,4 @@ target
 /*.iml
 /*.ipr
-/*.iws
+/*.iws
```
That's just minor, but you should agree with me that it is jut redundant noisy change.
The file has had a new line in the end before, but your IDE destroyed that and added the extra work for us :smile: 

As I said: minor. And will be fixed on merge. But just for the case you'd like to save some your time against my similar comments in the future.

pckeyan replied:
Got it

pckeyan replied:
Can I use the formatter at https://github.com/ndimiduk/spring-framework/blob/master/eclipse-code-formatter.xml for code formatting while contributing or let me know if you have a different location?

artembilan replied:
No, that's correct place for proper configs and rules.

artembilan replied:
This hasn't been fixed...
Don't worry: fixing on merge!

artembilan commented on src/main/resources/org/springframework/integration/aws/config/xml/spring-integration-aws-1.0.xsd:
```diff
@@ -65,9 +65,13 @@
+			<xsd:choice minOccurs="0" maxOccurs="2">
+				<xsd:element ref="integration:poller" minOccurs="0" maxOccurs="1"/>
+				<xsd:element name="request-handler-advice-chain" type="integration:handlerAdviceChainType" minOccurs="0" maxOccurs="1"/>
+			</xsd:choice>
```
That is correct change which really addresses the JIRA premise.
But I'm always open for more minor useful changes on the place. E.g. here I recommended you to take a look to the
`<xsd:attributeGroup ref="integration:channelAdapterAttributes"/>` for this `<s3-outbound-channel-adapter>` element. 

Let me know if it is a bit complex for you and we can raise a new JIRA for that would be some more practice for you how to do the proper contribution.

pckeyan replied:
I take that opportunity to work on new JIRA.

artembilan replied:
Here you are https://jira.spring.io/browse/INTEXT-197!

artembilan commented on src/test/java/org/springframework/integration/aws/s3/config/xml/AmazonS3OutboundChannelAdapterParserAWSTests.java:
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2002-2014 the original author or authors.
```
Since this is a new class, the date must be only `2015`.
Minor, of course, and will be addressed on the merge.

artembilan commented on src/test/java/org/springframework/integration/aws/s3/config/xml/AmazonS3OutboundChannelAdapterParserAWSTests.java:
```diff
@@ -0,0 +1,74 @@
+ * @since 1.0
+ *
+ */
+@Ignore
```
Is this ignored for the same reason as `DefaultAmazonS3OperationsAWSTests` ?
Can we `mock` it somehow to avoid any real AWS invocation, just because we have to test here only `adviceCalled` ?

Not sure that I will be able to address this on merge, so I need your input.

pckeyan replied:
Yes, I tried various ways to mock it but advice was not called. So I went back to this approach.

artembilan replied:
How about just throw an exception from the `FooAdvice` instead of `callback.execute()`?
In this case you don't have access to the adapter functionality, so no reason to worry about AWS stuff

artembilan commented on src/test/java/org/springframework/integration/aws/s3/config/xml/AmazonS3OutboundChannelAdapterParserAWSTests.java:
```diff
@@ -0,0 +1,74 @@
+
+
+	/**
+	 * Test case for the xml definition with a custom implementation of {@link AmazonS3Operations}
```
for me this comment doesn't reflect the reality.
Just remove it!

pckeyan replied:
Let me update the comment

artembilan commented on src/test/java/org/springframework/integration/aws/s3/config/xml/AmazonS3OutboundChannelAdapterParserAWSTests.java:
```diff
@@ -0,0 +1,74 @@
+	 * Tests the outbound channel adapter definition with a valid combination of attributes along with
+	 * request handler chain.
+	 *
+	 * Need AWS Connectivity.
```
Not sure why we need `AWS Connectivity` if we should just test that `request handler chain` works properly...

From other side I'd like do not have any *real* test in this project. Just because we build here just only *adapters* around AWS services and all hard work is done there we must only `verify` that we invoke correct AWS API.

Yes, there might be some tiny cases, which should be addressed in our adapters, when we don't have a high-level API from the AWS. But we will consider them separately.

Here we just deal with the `request handler chain`, so it is really bad to try to connect to the AWS for this our trivial test...

artembilan commented on src/test/resources/s3-valid-outbound-advice-cases.xml:
```diff
@@ -0,0 +1,42 @@
+	</bean>
+
+	<bean id="customOps"
+		class="org.springframework.integration.aws.s3.config.xml.AmazonS3DummyOperations"/>
```
Well, you use here the `mock`, but require the real AWS Connect...
Some inconsistency.

Please, fix the new line in the end of file bellow.

pckeyan replied:
I will fix it and commit it back with the mock.

artembilan replied:
Thank you!

artembilan commented on src/test/resources/s3-valid-outbound-advice-cases.xml:
```diff
@@ -0,0 +1,42 @@
+		<constructor-arg index="2" value="10"/>
+		<constructor-arg index="3" value="MILLISECONDS"/>
+		<constructor-arg index="4" ref="queue"/>
+	</bean>
```
Would you mind explaining the reason of this custom `ThreadPoolExecutor`?
Why isn't  `<task:executor>` just enough?

Oh! I see this is just copy/paste from existing tests.

Please, simplify the test config. Just because it doesn't relate to the test case and I don't see reason to pollute it with redundant code.

pckeyan commented:
INTEXT-187: Allow adding advice handler on AWS outbound channel adapter
JIRA: https://jira.spring.io/browse/INTEXT-187

* Modified the org/springframework/integration/aws/config/xml/spring-integration-aws-1.0.xsd to accomodate request-handler-advice-chain
	in s3 outbound adapter after feedback.
* Modified org.springframework.integration.aws.s3.config.xml.AmazonS3OutboundChannelAdapterParserTests - Added Unit Test Class with Mock Object to test the request handler chain.
* Modified s3-valid-outbound-cases.xml - Added separate definition withHandlerChain with mock Operations.

Please review at your convenience.

artembilan commented on src/test/java/org/springframework/integration/aws/s3/config/xml/AmazonS3OutboundChannelAdapterParserTests.java:
```diff
@@ -155,4 +172,13 @@ public String generateFileName(Message<?> message) {
+			return callback.execute();
+		}
+	}
+}
```
DITO...

I don't want to say "Bad!", but looks like you make me :smile:

pckeyan replied:
I used the code formatter from the link you provided earlier.

artembilan replied:
m-m-m. I think that formatter just doesn't provide the `New line in the end of file` option. 
But this is important. 
So, please, follow with my recommendations.

pckeyan replied:
Sure and Thanks a lot Artem. It was a great learning experience.

I will add a new line manually at the end of file.

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration-aws/commit/c674a8dace39f5a7adfd782034f50dec56a3b64c

@pckeyan , thank you for the contribution and patience.

Looking for more complex and more clear ones! :smile:
